File: typescript/22-most-frequent-char.ts
// Carácter más frecuente — Fácil · Strings
// En caso de empate, devuelve el carácter que aparece primero.
// Fuente: https://coding-challenges.dev/problems/typescript-caracter-mas-frecuente

export function mostFrequentChar(text: string): string {
  const counts = new Map<string, number>();
  let mostFrequent = "";
  let highestCount = 0;
  for (const char of text) {
    const count = (counts.get(char) ?? 0) + 1;
    counts.set(char, count);
    if (count > highestCount) {
      mostFrequent = char;
      highestCount = count;
    }
  }
  return mostFrequent;
}

File: python/22_most_frequent_char.py
"""Carácter más frecuente — Fácil · Strings

Devuelve el carácter más frecuente; en empate, el primero que aparece.
Los espacios también cuentan.

Fuente: https://coding-challenges.dev/problems/python-caracter-mas-frecuente
"""


def most_frequent_char(text: str) -> str:
    if not text:
        return ""
    counts = {}
    for char in text:
        counts[char] = counts.get(char, 0) + 1
    return max(counts, key=counts.get)
